### PR TITLE
Draft: Use HIP compiler builtins for atomics with memory scope/ordering

### PR DESCRIPTION
**Pull request by publixsubfan**

### CODE REVIEW

**Comment by dalg24** [comment]
Adding a reference to somewhat stalled effort to implement the same thing
https://github.com/kokkos/kokkos/pull/6031/files

**Comment by publixsubfan** [comment]
> Adding a reference to somewhat stalled effort to implement the same thing https://github.com/kokkos/kokkos/pull/6031/files

Oops, I didn't realize this was being worked on from the Kokkos side as well. Was there a blocker preventing it from being merged?

**Comment by dalg24** [comment]
> > Adding a reference to somewhat stalled effort to implement the same thing https://github.com/kokkos/kokkos/pull/6031/files
> 
> Oops, I didn't realize this was being worked on from the Kokkos side as well. Was there a blocker preventing it from being merged?

Some subtleties that needed more work.  I don't remember the detail on top of my head.
I think some uncertainty around how to implement the fences.  Also IIRC I was thinking that we might need to refactor how we do atomic compare exchange in the backends and do the specialization in terms of weak/strong compare exchange.

Maybe we can split things up because it is pretty clear how to do `atomic_fetch_op` part.

**Comment by dalg24** [comment]
Have you tried to deploy in RAJA?

**Comment by publixsubfan** [comment]
> Have you tried to deploy in RAJA or anywhere else?

I've done a cursory build/test with RAJA + this branch, but haven't done any testing with non-relaxed memory orderings yet. I think RAJA only assumes relaxed ordering and device-consistency with Desul at the moment.

Do you folks have litmus tests for stuff like acquire/release or sequential consistency in Kokkos?

**Comment by dalg24** [comment]
> Do you folks have litmus tests for stuff like acquire/release or sequential consistency in Kokkos?

We only have marginal use of the acquire/release semantics in our internals (e.g. thread fences in `UniqueToken`)

**Comment by trws** [comment]
Yeah, we're almost all relaxed at the moment, though it looks like we're going to care about the memory scopes more here pretty shortly.  This is actually extremely timely, and while I haven't done a detailed review it looks pretty clean.  Do you feel like this is in a good place for a full review and some testing?

File: atomics/include/desul/atomics/Fetch_Op_HIP.hpp [hunk 1]
```diff
@@ -9,9 +9,90 @@ SPDX-License-Identifier: (BSD-3-Clause)
...
+  return __hip_atomic_fetch_min(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+}
+
+template <typename T, typename MemoryOrder, typename MemoryScope>
+inline __device__ std::enable_if_t<std::is_arithmetic<T>::value, T>
+device_atomic_fetch_max(T* ptr, T val, MemoryOrder, MemoryScope) {
+  return __hip_atomic_fetch_max(
+      ptr, val, GCCMemoryOrder<MemoryOrder>::value, HIPMemoryScope<MemoryScope>::value);
+}
```
Comment (dalg24) [inline atomics/include/desul/atomics/Fetch_Op_HIP.hpp:69]: Did you test these guys?
I see that I had commented the fp overloads on my version https://github.com/kokkos/kokkos/pull/6031/files#diff-aae30b453c4ca28bccc8e0a5c46d3c1250b78cce5b3187ecb869cca34339e169R44-R45 which I suppose mean I had seen issues.  Of course I haven't taken notes and I don't remember what the deal was.

Comment (publixsubfan) [reply atomics/include/desul/atomics/Fetch_Op_HIP.hpp:69]: I ran through the `test-forall-atomic-basic-Hip.exe` in RAJA, built with `RAJA_ENABLE_EXHAUSTIVE=ON`.

Looking at the chunks of testing code [here](https://github.com/LLNL/RAJA/blob/a669daa06253d076e8ed57d35c55fd7a9e8c8614/test/include/RAJA_test-atomic-types.hpp#L24-L33) and [here](https://github.com/LLNL/RAJA/blob/a669daa06253d076e8ed57d35c55fd7a9e8c8614/test/functional/forall/atomic-basic/tests/test-forall-atomic-basic.hpp#L103-L112), running that test should exercise `atomicMin` and `atomicMax` for all types, including `float` and `double`.

**Comment by publixsubfan** [comment]
> Do you feel like this is in a good place for a full review and some testing?

I was hoping to add some ordered atomic tests to RAJA first before proceeding -- stuff like message passing or store buffering litmus tests to ensure that we're getting the correct behavior if using stronger orderings.

File: atomics/include/desul/atomics/Fetch_Op_HIP.hpp (hunk 1 above)
Comment (dalg24) [reply atomics/include/desul/atomics/Fetch_Op_HIP.hpp:69]: Please confirm what version(s) of ROCm you tested.

Comment (publixsubfan) [reply atomics/include/desul/atomics/Fetch_Op_HIP.hpp:69]: I tested 5.7.1 and 6.0.2. Are there other versions you'd want to test?

Comment (dalg24) [reply atomics/include/desul/atomics/Fetch_Op_HIP.hpp:69]: On the Kokkos end we need to support back to 5.2

**Comment by dalg24** [comment]
Closing in favor of #126 and #127

**Pull request closed without merging**